### PR TITLE
Frames in the constellation store LoadData rather than just URLs.

### LINKED ISSUE #14888: Save additional load data in Frame
After #14312, documents may be discarded and reloaded from the information stored in a `Frame` object. Currently we are just storing the URL, we should really also be storing the origin as well, so that when the document is reloaded, we can restore its origin. cc @jdm @cbrewster

### PR DESCRIPTION
**Pull request by asajeffrey**
Include LoadData in each session history entry, so that when we reload discarded documents we can include POST data etc.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14888
- [X] These changes do not require tests because we don't test document discarding.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15878)

### CODE REVIEW

File: components/constellation/frame.rs [hunk 1]
```diff
@@ -81,7 +80,7 @@ impl Frame {
     pub fn update_current(&mut self, pipeline_id: PipelineId, entry: &FrameState) {
         self.pipeline_id = pipeline_id;
         self.instant = entry.instant;
-        self.url = entry.url.clone();
+        self.load_data = entry.load_data.clone();
```
Comment (cbrewster) [inline components/constellation/frame.rs]: We should make `update_current` take ownership of `entry` to remove this clone.

File: components/constellation/constellation.rs [hunk 2]
```diff
@@ -2127,24 +2139,24 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
+            debug!("Replacing pipeline in existing frame with timestamp {:?}.", instant);
+            let entry = FrameState {
+                frame_id: frame_change.frame_id,
+                pipeline_id: Some(frame_change.new_pipeline_id),
+                load_data: frame_change.load_data.clone(),
+                instant: instant,
+            };
             self.traverse_to_entry(entry);
-            (evicted_id, false, None, false)
+            (None, false, None, false)
```
Comment (cbrewster) [inline components/constellation/constellation.rs]: Now that only one branch returns something other than `None` for `evicted_id` can we just move `close_pipeline` inside the only branch that uses it?

**Review by cbrewster — APPROVED** [review]
Looks good other than a couple nits

File: components/constellation/frame.rs (hunk 1 above)
Comment (asajeffrey) [inline components/constellation/frame.rs]: Done.

File: components/constellation/constellation.rs (hunk 2 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Afraid not:
```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
    --> /home/ajeffrey/github/asajeffrey/servo/components/constellation/constellation.rs:2165:17
     |
2156 |         } else if let Some(frame) = self.frames.get_mut(&frame_change.frame_id) {
     |                                     ----------- first mutable borrow occurs here
...
2165 |                 self.close_pipeline(evicted_id, DiscardBrowsingContext::No, ExitPipelineMode::Normal);
     |                 ^^^^ second mutable borrow occurs here
...
2171 |         };
     |         - first borrow ends here
```

Comment (cbrewster) [inline components/constellation/constellation.rs]: :(

Comment (asajeffrey) [inline components/constellation/constellation.rs]: Indeed. Is it still r=you?

File: components/constellation/frame.rs [hunk 3]
```diff
@@ -99,41 +98,29 @@ pub struct FrameState {
...
 
-    /// The URL for the document being loaded.
-    pub url: ServoUrl,
+    /// The data for the document being loaded.
+    pub load_data: LoadData,
 
     /// Is the new document replacing the current document (e.g. a reload)
     /// or pushing it into the session history (e.g. a navigation)?
-    pub replace: Option<FrameState>,
+    /// It it replacing an existing entry, we store its timestamp.
```
Comment (cbrewster) [inline components/constellation/frame.rs]: s/It it/If it is/

File: components/constellation/constellation.rs (hunk 2 above)
Comment (KiChjang) [inline components/constellation/constellation.rs]: Makes me want to have [2-phase borrowing](smallcultfollowing.com/babysteps/blog/2017/03/01/nested-method-calls-via-two-phase-borrowing/) here...

Comment (asajeffrey) [inline components/constellation/constellation.rs]: Indeed.

File: components/constellation/frame.rs (hunk 3 above)
Comment (asajeffrey) [inline components/constellation/frame.rs]: Fixed.

**Pull request merged**